### PR TITLE
Add ipcidr support

### PR DESCRIPTION
**Pull request by MarcoPolo**
Adds the `ipcidr` protocol.

Example: `/ip4/1.2.3.0/ipcidr/24`

Can be used by ipv6 as well

### CODE REVIEW

File: net/convert_test.go [hunk 1]
```diff
@@ -202,3 +202,40 @@ func TestDialArgs(t *testing.T) {
...
+		contained []bool
+	}
+
+	testCases := []testCase{
+		{
+			ma:        "/ip4/1.2.3.0/ipcidr/24",
+			ips:       []string{"1.2.3.4", "1.2.3.9", "2.1.1.1"},
+			contained: []bool{true, true, false},
+		},
+	}
```
Comment (marten-seemann) [inline net/convert_test.go:220]: Can we have test cases for the error conditions as well?

File: transcoders.go [hunk 2]
```diff
@@ -54,6 +54,15 @@ func (t twrp) ValidateBytes(b []byte) error {
 var TranscoderIP4 = NewTranscoderFromFunctions(ip4StB, ip4BtS, nil)
 var TranscoderIP6 = NewTranscoderFromFunctions(ip6StB, ip6BtS, nil)
 var TranscoderIP6Zone = NewTranscoderFromFunctions(ip6zoneStB, ip6zoneBtS, ip6zoneVal)
+var TranscoderIPCIDR = NewTranscoderFromFunctions(func(s string) ([]byte, error) {
```
Comment (marten-seemann) [inline transcoders.go:57]: Looks like the convention here was to define separate functions.

File: multiaddr_test.go [hunk 3]
```diff
@@ -101,9 +101,11 @@ func TestConstructSucceeds(t *testing.T) {
 	cases := []string{
 		"/ip4/1.2.3.4",
 		"/ip4/0.0.0.0",
+		"/ip4/192.0.2.0/ipcidr/24",
```
Comment (marten-seemann) [inline multiaddr_test.go:104]: Can you also add a few entries to `TestConstructFails`?

File: transcoders.go (hunk 2 above)
Comment (MarcoPolo) [reply transcoders.go:57]: Do you know if there's a specific reason?

File: transcoders.go [hunk 4]
```diff
@@ -54,6 +54,19 @@ func (t twrp) ValidateBytes(b []byte) error {
 var TranscoderIP4 = NewTranscoderFromFunctions(ip4StB, ip4BtS, nil)
 var TranscoderIP6 = NewTranscoderFromFunctions(ip6StB, ip6BtS, nil)
 var TranscoderIP6Zone = NewTranscoderFromFunctions(ip6zoneStB, ip6zoneBtS, ip6zoneVal)
+var TranscoderIPCIDR = NewTranscoderFromFunctions(ipcidrStB, ipcidrBtS, nil)
+
+func ipcidrBtS(b []byte) (string, error) {
+	return strconv.Itoa(int(b[0])), nil
```
Comment (marten-seemann) [inline transcoders.go:60]: This probably should have a length check to be extra careful (like in `ip6zoneStB`).

**Review by marten-seemann — APPROVED** [review]
LGTM, module one defensive check.

**Comment by MarcoPolo** [comment]
@marten-seemann can you merge please? I don’t have access

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**